### PR TITLE
added e2e tests for create, update applicant

### PR DESCRIPTION
**Pull request by maheshKRex**
https://github.com/EmpowerPlastic/empowerchain/issues/408

### CODE REVIEW

File: chain/tests/e2e/plasticcredit/applicant_tx_test.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	s.Require().NoError(err)
+
+	testCases := map[string]struct {
+		args              []string
+		expectedErrOnSend bool
+		expectedErrOnExec bool
+		expectedErrMsg    string
+		expectedState     *plasticcredit.Applicant
+	}{
+		"create applicant": {
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/applicant_tx_test.go:26]: We could add the test for empty name (error) and empty description (which should pass) and invalid admin address.

File: chain/tests/e2e/plasticcredit/project_tx_test.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+	notApplicantKey, _ := val.ClientCtx.Keyring.Key(issuerKey)
+	testCases := map[string]struct {
+		args              []string
+		expectedErrOnSend bool
+		expectedErrOnExec bool
+		expectedErrMsg    string
+		expectedState     *plasticcredit.Project
+	}{
+
+		"create project": {
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/project_tx_test.go:26]: Empty name and empty description?

File: chain/tests/e2e/plasticcredit/project_tx_test.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+	s.Require().NoError(err)
+	testCases := map[string]struct {
+		args              []string
+		expectedErrOnSend bool
+		expectedErrOnExec bool
+		expectedErrMsg    string
+		expectedState     *plasticcredit.Project
+	}{
+
+		"approve project": {
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/project_tx_test.go:156]:
```suggestion
		"approve new project": {
```
Just for clarification

File: chain/tests/e2e/plasticcredit/project_tx_test.go [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+			nil,
+		},
+		"invalid admin": {
+			[]string{"4", fmt.Sprintf("--%s=%s", flags.FlagFrom, notAdminKey.Name)},
+			false,
+			true,
+			"unauthorized",
+			nil,
+		},
+		"project already approved": {
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/project_tx_test.go:196]: Maybe let's add approving rejected project also?

File: chain/tests/e2e/plasticcredit/project_tx_test.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+			nil,
+		},
+		"reject non-existing project": {
+			[]string{"62", fmt.Sprintf("--%s=%s", flags.FlagFrom, issuerKey.Name)},
+			false,
+			true,
+			"project not found",
+			nil,
+		},
+		"project already approved": {
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/project_tx_test.go:266]: Rejecting already rejected and suspended?

File: chain/tests/e2e/plasticcredit/project_tx_test.go [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+	notAdminKey, err := val.ClientCtx.Keyring.Key(applicantKey)
+	s.Require().NoError(err)
+	testCases := map[string]struct {
+		args              []string
+		expectedErrOnSend bool
+		expectedErrOnExec bool
+		expectedErrMsg    string
+		expectedState     *plasticcredit.Project
+	}{
+		"reject project": {
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/project_tx_test.go:239]:
```suggestion
		"reject new project": {
```

File: chain/tests/e2e/plasticcredit/project_tx_test.go [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+	notAdminKey, err := val.ClientCtx.Keyring.Key(applicantKey)
+	s.Require().NoError(err)
+	testCases := map[string]struct {
+		args              []string
+		expectedErrOnSend bool
+		expectedErrOnExec bool
+		expectedErrMsg    string
+		expectedState     *plasticcredit.Project
+	}{
+		"suspend project": {
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/project_tx_test.go:309]:
```suggestion
		"suspend approved project": {
```

File: chain/tests/e2e/plasticcredit/project_tx_test.go [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+			nil,
+		},
+		"suspending non-existing project": {
+			[]string{"62", fmt.Sprintf("--%s=%s", flags.FlagFrom, issuerKey.Name)},
+			false,
+			true,
+			"project not found",
+			nil,
+		},
+		"project already rejected": {
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/project_tx_test.go:336]: Suspending already suspended and new projects?

File: chain/tests/e2e/plasticcredit/project_tx_test.go [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+		expectedErrMsg    string
+		expectedState     *plasticcredit.Project
+	}{
+
+		"update project": {
+			[]string{"3", "My Updated Project", fmt.Sprintf("--%s=%s", flags.FlagFrom, applicantKey.Name)},
+			false,
+			false,
+			"",
+			&plasticcredit.Project{
```
Comment (ba1ciu) [inline chain/tests/e2e/plasticcredit/project_tx_test.go:90]: Project with ID 3 is used for both update and approve testing. You should use different for each case. We should keep in mind that the test should run successfully both as a whole suite and independently. If you run just this test, it will fail, because Project 3 is initially in new state.

**Review by ba1ciu — CHANGES_REQUESTED** [review]
Couple suggestions about other test scenarios, mostly testing all possible transitions between project statuses and one small issue with project update to fix. Those are not specifically for the applicant case, so we can create separate issue for those or fix right away :)

File: chain/tests/e2e/plasticcredit/applicant_tx_test.go (hunk 1 above)
Comment (maheshKRex) [reply chain/tests/e2e/plasticcredit/applicant_tx_test.go:26]: @ba1ciu i've been struggling with the invalid admin test for create applicant. Similarly to other invalid admin tests i tried to provide a test case with simply an other address string as the admin account address instead of issuer/applicant address from keyring. 
`"invalid admin address": {
			[]string{"empower1qnk2n4nlkpw9xfqntladh74w6ujtulwnz7rf", "New", "Grab that bottle", fmt.Sprintf("--%s=%s", flags.FlagFrom,issuerKey.Name)},
		`
But the clitestutil.ExecTestCLICmd fails to pass this string to the test with error key not found, so when invoking the CLI this way, does is always need to have a valid key AccAddress, like issuer / applicant key address?
`empower1qnk2n4nlkpw9xfqntladh74w6ujtulwnz7rf8.info: key not found`

Comment (maheshKRex) [reply chain/tests/e2e/plasticcredit/applicant_tx_test.go:26]: found that this is imposed by validation check at CLI x/plasticcredit/client/cli/tx.go forcing to take --from addr as the Admin, making it to always have a valid AccAddress. So it can be any address as we don't restrict anybody with a signing address to go ahead and create an applicant.
`if err := cmd.Flags().Set(flags.FlagFrom, args[0]); err != nil {
	return err
}`
So should we still do an invalid admin test, or just test for validity of from address.

File: chain/tests/e2e/plasticcredit/project_tx_test.go (hunk 2 above)
Comment (maheshKRex) [reply chain/tests/e2e/plasticcredit/project_tx_test.go]: added.

File: chain/tests/e2e/plasticcredit/project_tx_test.go (hunk 9 above)
Comment (maheshKRex) [reply chain/tests/e2e/plasticcredit/project_tx_test.go:103]: agree. updated to use new project for this. https://github.com/EmpowerPlastic/empowerchain/blob/702dbf9b8b0940edd69e170941155eba76c195ba/chain/tests/e2e/plasticcredit/project_tx_test.go#L104

File: chain/tests/e2e/plasticcredit/project_tx_test.go (hunk 8 above)
Comment (maheshKRex) [reply chain/tests/e2e/plasticcredit/project_tx_test.go]: added.

File: chain/tests/e2e/plasticcredit/project_tx_test.go (hunk 5 above)
Comment (maheshKRex) [reply chain/tests/e2e/plasticcredit/project_tx_test.go]: added.

File: chain/tests/e2e/plasticcredit/applicant_tx_test.go (hunk 1 above)
Comment (ba1ciu) [reply chain/tests/e2e/plasticcredit/applicant_tx_test.go]: I think we should leave validation in tx.go as it is, so the cli works both with addresses and key names.
If the message signer is an admin and therefore you cannot give anyone other than yourself an admin role, then I suggest we either drop the admin param and take it from the --from flag or allow to assign admin to others by adding a creator field, which will hold the address from --from. Wdyt, @gjermundgaraba?

Comment (maheshKRex) [reply chain/tests/e2e/plasticcredit/applicant_tx_test.go]: yes, updated create-applicant to take admin using --from flag, and it looks much neater.

**Review by ba1ciu — APPROVED** [review]
